PureScript Community Code of Conduct (#13)
purescript/governance: CODE_OF_CONDUCT.md

# PureScript Community Code of Conduct

Our community is dedicated to helping people learn and use the PureScript programming language. Our members can't do this if they are being harassed, stalked, or discriminated against. It's also hard to do that if you're regularly made to experience emotionally draining situations or made to wonder whether you belong. Accordingly, all PureScript community members are expected to show respect and courtesy towards each other in all interactions in our community.

We need a common understanding of what it means to "show respect and courtesy towards each other" and so we have adopted the following code of conduct. This code of conduct is enforced by our moderators and applies in all community spaces.

## Our Purpose

The PureScript community spaces are places to discuss the PureScript programming language. In these spaces:

* We talk about the language, its related libraries and infrastructure, and our experiences using it.
* We help PureScript programmers learn the language and we ask for help becoming stronger programmers ourselves. 

> * We discuss general functional programming concepts and other parts of programming, type theory, and mathematics that are relevant to Purescript.

Reviewer: ```suggestion
* We discuss general functional programming concepts and other parts of programming, type theory, and mathematics that are relevant to PureScript.
```

- **Maliciousness towards other members**: deliberately attempting to make others feel bad, name-calling, or singling others out for derision. For example: telling someone they're a poor programmer.
- **Being especially unpleasant**: for example, if we've received several reports of annoying, rude, or especially distracting behavior.

The rule around prejudice and blatant -isms is especially complicated because not everyone agrees on what constitutes a violation. Members of the PureScript community live in disparate cultures and speak different languages. Harmful prejudices are baked into society in ways that can make them hard to recognize, and not everyone experiences prejudice in the same way: subtle ageism won't hurt someone who is young in the same way it hurts someone who is older.

> If you are told that a symbol, phrase, or message is unwelcome, please apologize and move on. It's not a big deal for community members to make a mistake from time to time.

Reviewer: ‘If you are told’ by a moderator? By another member of the community? This is an important question; it's the difference between a community where unacceptable language is determined by the judgment of the moderators (and limited by their point-in-time awareness of the potential harm done by language), and a community where unacceptable language is defined as the union of everything found unacceptable by each member of the community.

If the latter is what is intended, you might want to rethink the language above about ‘arguing with someone who has pointed that out to you’. If the community norms get determined by the community, but the moderators don't permit the community to productively debate those norms, that would seem to create a sort of ratchet effect where it's trivial to establish new forbidden language and impossible to push back in the other direction, and I assume that's not how you actually intend to run things.

Just to emphasize, none of this is meant to be feedback about how you should and shouldn't moderate; I don't consider myself qualified to give that feedback. My primary concern here is that you clearly communicate your vision as moderators, with the intended audience being someone who would like to know in advance what they can expect before they invest in the community. My position is that a CoC ought to give as much detail as is practical on potentially thorny questions like this, not in order to bind the moderators into a legalistic framework, but rather in order to dispel ambiguity in the minds of the hesitant.

Author: > ‘If you are told’ by a moderator

This is the correct reading. I'd assume that in practice, for particularly obvious cases, members of the community might note some language or action isn't generally considered acceptable here. But it's only when you are told _by a moderator_ that the rule about not arguing comes into play.

There is no intent to empower individuals to arbitrarily decide on new forbidden language or behavior, if only because it then becomes extremely difficult to moderate. But individuals can certainly suggest or ask for certain things to be moderated, in which case it is still up to the moderators to make any 'official' decision on it.

Author: I've updated this! Thanks for pointing it out.

## Enforcement

In the previous section we categorized unacceptable behavior into abuse and unwelcoming behavior. Moderators will respond differently depending on the type of violation that occurs.

> If we witness or receive a report about abusive behavior, we will contact the perpetrator to verify what has transpired, and they will be removed from the PureScript community. Their Discord and Discourse accounts will be deactivated and they will be removed from PureScript GitHub organizations.

Reviewer: ‘[C]ontact the perpetrator to verify what has transpired’ sounds like maybe the ‘perpetrator’ (alleged perpetrator? accused? complainee?) has a chance to give their side of the story and talk their way out of the following consequences, but the rest of the language of this paragraph doesn't imply that it's conditional on the report being verified. Here are a few example edits that I think would clarify this, depending on where you intend to fall on the believe-all-reports/benefit-of-the-doubt spectrum:

```suggestion
If we witness or receive a report about abusive behavior, we will contact the perpetrator to inform them of the observation or report. They will be removed from the PureScript community. Their Discord and Discourse accounts will be deactivated and they will be removed from PureScript GitHub organizations.
```

```suggestion
If we witness or receive a report about abusive behavior, we will contact the complainee to verify what has transpired. At the time of the report, the complainee will be removed from the PureScript community. Their Discord and Discourse accounts will be deactivated and they will be removed from PureScript GitHub organizations. If the complainee contests the facts of the report, we will investigate further and, if in the judgment of the moderation team no abusive behavior took place, the complainee's accounts will be reinstated with no further consequences.
```

```suggestion
If we witness or receive a report about abusive behavior, we will contact the complainee and investigate the details of the incident. In the event that the incident constitutes abusive behavior in the opinion of the moderation team, the perpetrator will be removed from the PureScript community. Their Discord and Discourse accounts will be deactivated and they will be removed from PureScript GitHub organizations.
```

# PureScript Community Code of Conduct

> The PureScript community follows some simple ground rules to protect safe, inclusive spaces to discuss the PureScript programming language.

Reviewer 2: Should the comma between "safe" and "inclusive" be replaced with an "and"? And perhaps replace "to" with "where we can". So would read "The PureScript community follows some simple ground rules to protect safe and inclusive spaces where we can discuss the PureScript programming language."

Reviewer 3: I think either rendering (Thomas' current one and your suggested one) still say the same thing. So, I don't really have an opinion here.

All community members are expected to act and interact in ways that contribute to an open, welcoming, diverse, inclusive, and healthy community.

## What's our purpose?

The PureScript community channels are places to discuss the PureScript programming language. 

* We talk about the language, its related libraries and infrastructure, and our experiences using it.

> * We help PureScript programmers learn the language, and we ask for help becoming stronger programmers ourselves. 

Reviewer 2: I think the comma before the "and" should be removed. Perhaps this is a writing style thing but all the other commas have some grammatical purpose while this one appears to indicate a pause.

Reviewer 3: The comma here could be removed or kept as-is. Again, I don't have an opinion.

* We discuss general functional programming concepts and other parts of programming, type theory, and mathematics that are relevant to Purescript.
* We organize open-source work, meetups, hackathons, and other community events.

> * We discuss what we can do to grow an inclusive, friendly, respectful community.

Reviewer 2: Kinda trivial but all your other lists end with an "and". So perhaps this should be "We discuss what we can do to grow an inclusive, friendly, and respectful community."

Reviewer 3: I think this worth adding. It does make things a bit more readable.

## What is okay?

Examples of behavior that contributes to a positive environment for our community include:

* Demonstrating empathy and kindness toward other people.
* Being respectful of differing opinions, viewpoints, and experiences.
* Giving and gracefully accepting constructive feedback.
* Accepting responsibility and apologizing to those affected by our mistakes, and learning from the experience.

> As participants in the PureScript community, we pledge to make participation in our community a harassment-free experience for everyone -- regardless of age, body size, disability, ethnicity, sex characteristics, gender identity and expression, level of experience, education, socio-economic status, nationality, personal appearance, race, caste, color, religion, or sexual identity and orientation.

Reviewer 2: Split "gender identity and expression" into "gender identity, gender expression". Similarly "religion, or sexual identity and orientation" becomes "religion, sexual identity or sexual orientation."

Reviewer 3: I think those are intentionally grouped together.

Reviewer 2: I think that they probably were originally grouped together intentionally. But I believe splitting them is worthwhile for two reasons.
1. By grouping "gender identity and expression" and "sexual identity and orientation" we are saying that identity and expression in those instances are tied together in a way that, for example, race and color are not. That may be true -- it's a long way outside my field of expertise. But what we are doing with this sentence is highlighting examples and I think it would be better to be silent on the relationships between any of these attributes. We might also want to sort the list in alphabetical order to remove any hint of hierarchy between them.

2. No reasonable person would make the argument that "I was only harassing this person based on their sexual orientation, not their sexual identity. Therefore I was not harassing them based on their "sexual identity and sexual orientation" because 'false and true equals false'". But if there ever was a person who would make that argument it would be a troll with a background in computer science.

Reviewer 4: I think this list should be removed altogether. You're never gonna get it right and it's logically the same statement if you remove it: "we pledge to make participation in our community a harassment-free experience for everyone."

Reviewer 5: Strongly disagree. It may be the case that it’ll never be perfect but we can always amend it in the future if we need to. To me, if a CoC doesn’t actually list the protected characteristics like this, it’s a good signal that the community doesn’t recognise one or more of them as being worthy of protection, and this would make me _less_ likely to want to participate.

Reviewer 4: @hdgarrood but doesn't saying "harassment-free" mean no harassment of any kind? Sans the list, you think people will read this and think to themselves "hm, perhaps I am the unstated exception to this rule"?

Reviewer: It's that different people have different definitions of what constitutes harassment. A list like this signals that the moderators are (or at least claim to be) aware of the specific concerns of the mentioned communities, and intend to take them into account when determining whether behavior merits moderator action. Sans the list, people will read this and wonder whether ‘harassment-free’ means that the specific, possibly controversial form of harassment that they've experienced or seen others experience is going to get a pass because it's not an overt threat.

Reviewer 4: Okay, this is a very good point. I still don't like the list as is, because to me it logically disguises it's true purpose: to more clearly define what the community considers harassment. Personally, I would prefer it be reworded/reorganized to be inline with this purpose.

Author: I've tried to reword things so that this sentiment (define some examples of what the community considers harassment) is more obvious.

## What is not okay?

Examples of unacceptable behavior include:

* The use of sexualized language or imagery, and sexual attention or advances of any kind.
* Trolling, insulting or derogatory comments, and personal or political attacks.
* Public or private harassment.
* Publishing others’ private information, such as a physical or email address, without their explicit permission.
* Other conduct which could reasonably be considered inappropriate in a professional setting.

> 

Reviewer 6: Is there a thematic breakup here to have 2 newlines? Typo or `---`?

Author: This was unintentional -- it's just meant to be one newline.

## Where does this Code of Conduct apply?

This code of conduct covers anyone participating in PureScript community channels, including our Discord server, Discourse forum, and GitHub properties. For members of the PureScript team, it applies anywhere they are representing the PureScript project (for example, on social media).

## How do members report violations?

You can report abusive, harassing, or otherwise unacceptable behavior to the community moderators team or to a particular moderator if you are not comfortable sharing your concern with the full team.

All complaints will be reviewed and investigated promptly and fairly. Your communication with moderators will not be shared with anyone else without your permission.

Here are per-platform instructions:

> * Discord: Send a private message to the `@Moderators` team. To find a specific moderator to message, use the members sidebar under the `Moderators` role to find their profile.

Reviewer 5: It appears that this only works for moderators who are currently available; I can't find a way of listing all the members of the moderators role including those who are not currently available.

Author: That's true, though you can still send a message to the group. We could have a brief meeting with current moderators and ask if everyone is OK with their name and email address being included in the document, or we could set up a `moderators@purescript.org` email address that forwards to moderator inboxes.

> * Discourse: Send a private message to the `moderators` group. To find a specific moderator to message, see the `moderators` group page: https://discourse.purescript.org/g/moderators

Reviewer: This group page doesn't exist or is private, says Discourse.

Reviewer 5: It works for me, so it must only be visible to moderators. This is part of why I would prefer we list moderators explicitly with their names and email addresses here - I think making it easy to report incidents should be the priority, and listing email addresses seems like the easiest and most foolproof way to ensure that people can report incidents. Requiring people to navigate or sign up for some service they're not familiar with or for which the instructions might not work could be a very poor experience if you're already upset because of the incident you're trying to report, and not everyone is on Discord or Discourse.

Author: I did find that moderators can be found directly via the `about` page, which applies for everyone regardless of being logged in:
https://discourse.purescript.org/about

* All Other: If an event occurs somewhere else, please reach out to moderators on either Discord or Discourse.

## How will you enforce the rules?

> Our community leaders follow the below guidelines to determine the consequences for any action they deem in violation of this Code of Conduct:

Reviewer 2: Replace "community leaders" with "moderators". Or define what we mean by "community leader".

Reviewer 3: I think "moderators" would be clearer. However, "community leaders" sounds more personal. Perhaps we should clarify how one can identify a "community leader" in the chatroom?

1. **Correction**

   Cause: Use of inappropriate language or other behavior deemed unprofessional or unwelcome in the community.

>    Consequence: A private, written warning from community leaders, providing clarity around the nature of the violation and an explanation of why the behavior was inappropriate. A public apology may be requested.

Author: We may also want to include that moderators can edit or delete a message, or ask that the person edits or deletes their message, if it's deemed something that shouldn't be preserved in the channel. This is a sensitive line to walk, but I suspect it will sometimes be necessary.

Reviewer 2: I agree. Also, replace "community leaders" with "the moderators", remove the comma after leaders, replace "providing clarity around" with "clarifying", replace "an explanation of" with "explaining". 

So it now reads "A private, written warning from moderators clarifying the nature of the violation and explaining why the behavior was inappropriate."

2. **Warning**

   Cause: A violation through a single incident or series of actions.

   Consequence: A warning with consequences for continued behavior. No interaction with the people involved, including unsolicited interaction with those enforcing the Code of Conduct, for a specified period of time. Violating these terms may lead to a temporary or permanent ban.

3. **Temporary Ban**

>    Cause: A serious violation of community standards, including sustained inappropriate behavior.

Reviewer 2: I think the Causes should be in the same general format so people can see the escalation. So change to "A violation through a single serious incident or sustained inappropriate behavior".

   Consequence: A temporary ban from any sort of interaction or public communication with the community for a specified period of time. No interaction with the people involved, including unsolicited interaction with those enforcing the Code of Conduct, is allowed during this period. Violating these terms may lead to a permanent ban.

4. **Permanent Ban**

>    Cause: Demonstrating a pattern of violation of community standards, including sustained inappropriate behavior, harassment of an individual, or aggression toward or disparagement of classes of individuals.

Reviewer 2: I think it's important to make it clear that a permanent ban can be handed down for a single violation if that violation is serious enough. So change to "A violation through a single very serious incident or anything that demonstrates a pattern of violation of community standards".

Reviewer 3: This is a good point. Ideally, most issues would go through levels of escalation and hopefully be addressed in lower-levels. However, it is possible that we could get a troll whose intent is to do harm. We wouldn't want to use this power frequently, but should still have it available if such a troll came.

Reviewer 2: I agree. Is the escalation point worth making explicitly? I think it comes across fairly well just from the structure of the section.

We need a common understanding of what it means to "show respect and courtesy towards each other" and so we have adopted the following code of conduct. This code of conduct is enforced by our moderators and applies in all community spaces.

## Our Purpose

The PureScript community spaces are places to discuss the PureScript programming language. In these spaces:

* We talk about the language, its related libraries and infrastructure, and our experiences using it.
* We help PureScript programmers learn the language, and we ask for help becoming stronger programmers ourselves. 
* We discuss general functional programming concepts and other parts of programming, type theory, and mathematics that are relevant to Purescript.
* We organize open-source work, meetups, hackathons, and other community events.
* We discuss what we can do to grow an inclusive, friendly, respectful community.

## Unacceptable Behavior

The following types of behavior are unacceptable in the PureScript community and constitute violations of this code of conduct.

> This code of conduct is a set of guidelines that helps our community protect safe, inclusive spaces to discuss PureScript. However, it doesn't enumerate all possible forms of unwelcome behavior, and incidents are evaluated on a case-by-case basis so as to understand their full context.

Reviewer 3: The line above says, "The following types of behavior are unacceptable..."

And now this line says "... it doesn't enumerate all possible forms of unwelcome behavior.."

The first line originally had a list that was deleted, so was the first line supposed to be deleted as well?

Author: I could do a better job merging them together. The point I'm trying to make is that this is a set of guidelines, but it is ultimately up to the discretion of moderators to decide what is acceptable even if it's not technically written out here. But perhaps this phrasing is verbose and unnecessary.

Author: On a second look, I could just swap the two paragraphs
